runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/LockFreeHistogram.java

```diff
@@ -0,0 +1,216 @@
+    abstract double sum();
+
+    public abstract long count();
+
+    public static final OutlierStatistic EMPTY = create(0, 0);
+
+    public static OutlierStatistic create(double sum, long count) {
+      return new AutoValue_LockFreeHistogram_OutlierStatistic(sum, count);
+    }
+
+    public OutlierStatistic combine(double value) {
+      return create(sum() + value, count() + 1);
+    }
+
+    public double mean() {
+      if (count() == 0) {
+        return 0;
+      }
+      return sum() / count();
+    }
+  }
+
+  /**
+   * The snapshot of a histogram. The snapshot contains the overflow/underflow statistic, number of
+   * values recorded in each bucket, and the BucketType of the underlying histogram.
+   */
+  @AutoValue
+  public abstract static class Snapshot {
+    public abstract OutlierStatistic underflowStatistic();
+
+    public abstract OutlierStatistic overflowStatistic();
+
+    public abstract ImmutableLongArray buckets();
+
+    public abstract HistogramData.BucketType bucketType();
+
+    public static Snapshot create(
+        OutlierStatistic underflowStatistic,
+        OutlierStatistic overflowStatistic,
+        ImmutableLongArray buckets,
+        HistogramData.BucketType bucketType) {
+      return new AutoValue_LockFreeHistogram_Snapshot(
+          underflowStatistic, overflowStatistic, buckets, bucketType);
+    }
+
+    @Memoized
+    public long totalCount() {
+      long count = 0;
+      count += underflowStatistic().count();
+      count += overflowStatistic().count();
+      count += buckets().stream().sum();
+
+      return count;
+    }
+  }
+
+  /**
+   * Extract a delta update of this histogram. Update represents values that have been recorded in
+   * this histogram since the last time this method was called.
+   *
+   * <p>If this histogram is being updated concurrent to this method, then the returned snapshot is
+   * not guarenteed to contain those updates. However, those updates are not dropped and will be
+   * represented in a future call to this method.
+   *
+   * <p>If this histogram has not been updated since the last call to this method, an empty optional
+   * is returned.
+   */
+  public Optional<Snapshot> getSnapshotAndReset() {
+    if (!dirty.getAndSet(false)) {
+      return Optional.empty();
+    }
+
+    ImmutableLongArray.Builder bucketsSnapshotBuilder =
+        ImmutableLongArray.builder(buckets.length());
+    for (int i = 0; i < buckets.length(); i++) {
+      bucketsSnapshotBuilder.add(buckets.getAndSet(i, 0));
+    }
+    OutlierStatistic overflowSnapshot = overflowStatistic.getAndSet(OutlierStatistic.EMPTY);
+    OutlierStatistic underflowSnapshot = underflowStatistic.getAndSet(OutlierStatistic.EMPTY);
+
+    return Optional.of(
+        Snapshot.create(
+            underflowSnapshot, overflowSnapshot, bucketsSnapshotBuilder.build(), bucketType));
+  }
+
+  /** Create a histogram. */
+  public LockFreeHistogram(KV<MetricName, HistogramData.BucketType> kv) {
```

Comment:
move constructor under members for readability.

nit: (might be my personal style preference, so up to you) but also initialize all members in constructor instead of where declared for readability